mkt/extensions/models.py

```diff
@@ -179,26 +120,165 @@ def mini_manifest(self):
         # version. We don't bother with locales at the moment, this probably
         # breaks extensions using https://developer.chrome.com/extensions/i18n
         # but we'll deal with that later.
+        try:
+            version = self.latest_public_version
+        except ExtensionVersion.DoesNotExist:
+            return {}
         mini_manifest = {
-            'name': self.manifest['name'],
-            'package_path': self.download_url,
-            'version': self.manifest['version']
+            'name': version.manifest['name'],
+            'package_path': version.download_url,
+            'version': version.manifest['version']
         }
-        if 'author' in self.manifest:
+        if 'author' in version.manifest:
             mini_manifest['developer'] = {
-                'name': self.manifest['author']
+                'name': version.manifest['author']
             }
-        if 'description' in self.manifest:
-            mini_manifest['description'] = self.manifest['description']
+        if 'description' in version.manifest:
+            mini_manifest['description'] = version.manifest['description']
         return mini_manifest
 
+    @property
+    def mini_manifest_url(self):
+        return absolutify(reverse('extension.mini_manifest',
+                                  kwargs={'uuid': self.uuid}))
+
+    def save(self, *args, **kwargs):
+        if not self.slug:
+            self.clean_slug()
+        return super(Extension, self).save(*args, **kwargs)
+
+    def __unicode__(self):
+        return u'%s: %s' % (self.pk, self.name)
+
+    def update_status_according_to_versions(self):
+        """Update `status`, `latest_version` and `latest_public_version`
+        properties depending on the `status` on the ExtensionVersion
+        instances attached to this Extension."""
+        # If there is a public version available, the extension should be
+        # public. If not, and if there is a pending version available, it
+        # should be pending. Otherwise it should just be incomplete.
+        if self.versions.filter(status=STATUS_PUBLIC).exists():
+            self.update(status=STATUS_PUBLIC)
+        elif self.versions.filter(status=STATUS_PENDING).exists():
+            self.update(status=STATUS_PENDING)
```

Comment:
would this mean the add-on itself would become de-listed? i figure it should be possible to upload a new version of an add-on, keeping the add-on public, but just have a pending version.